<a href="https://colab.research.google.com/github/puttipongyy/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [18]:
import pandas as pd
import os

In [19]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [20]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [21]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [22]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

Quiz: ข้อมูลมาจากกี่ประเทศ

In [23]:
len(set(data['Country']))

38

In [24]:
data['Country'].unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

# แปลง data ให้อยู่ในรูปแบบ Transaction data ด้วย *groupby.apply(list)*

In [25]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [26]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [27]:
!pip install apyori

In [28]:
from apyori import apriori

In [29]:
apriori(transactions,min_support = 0.5)

<generator object apriori at 0x7bce3d8b2030>

In [30]:
list(apriori(transactions,min_support = 0.065))

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [31]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [32]:
len(frequentItemsets)

4

In [33]:
frequentItemsets[0]

RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)])

In [34]:
frequentItemsets[0][0]

frozenset({'22423'})

In [35]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [36]:
frequentItemsets[0][0].__len__()

1

In [37]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [38]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

Quiz <br>
stock code 22386 and 85099B คือสินค้าอะไร

In [39]:
set(data[(data['StockCode']=='22386')]['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [40]:
set(data[(data['StockCode']=='85099B')]['Description'])

{'JUMBO BAG RED RETROSPOT'}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),
<br>
<br>

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [41]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'548373',
 '543486',
 'C537853',
 '558540',
 '559827',
 '562377',
 '559295',
 '552662',
 '549951',
 '539468',
 '567886',
 '550470',
 '547580',
 '565994',
 '578949',
 '546315',
 '549425',
 '550638',
 '538177',
 '574856',
 '566603',
 '566981',
 '558918',
 '563613',
 '542622',
 '571713',
 '546545',
 '562843',
 '568134',
 'C566167',
 '550968',
 '553901',
 '543399',
 '564972',
 '572550',
 '565234',
 '576644',
 '567140',
 '569918',
 '552694',
 '565424',
 '552642',
 '555276',
 '542583',
 '563505',
 '540499',
 '543201',
 '553013',
 '551014',
 '542527',
 '575063',
 '567336',
 'C539071',
 'C574507',
 '536523',
 '562282',
 '575703',
 '537887',
 '558864',
 '559357',
 '568346',
 '577076',
 '546096',
 '545340',
 '572854',
 '560580',
 '562711',
 '573320',
 '579102',
 '558475',
 '579196',
 '560790',
 '577512',
 '543045',
 '547107',
 '543732',
 '550986',
 '540372',
 '537666',
 '556259',
 '581170',
 '542664',
 '550921',
 '544657',
 '580173',
 '536530',
 'C578144',
 '578936',
 '546008',
 '575176',
 '536

In [42]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

#HW 1 หา Frequent Item Set และ Association Rules
* ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลา ตรุษจีน ในทวีป Asia  (15 Jan - 15 Feb)
* ของช่วงเวลา  Christmas & New Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* โดยจัดกลุ่มตามประเทศ + เดือน


In [43]:
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country  
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  
4  12/01/2010 08:26       3.39     17850.0  United Kingdom

In [44]:
data_HW = data.copy()
data_HW

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [45]:
data_dadate = data.copy()

In [46]:

# Given data
countries = ['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
             'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
             'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
             'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
             'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
             'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
             'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
             'European Community', 'Malta', 'RSA']

# Fictional mapping of countries to continents (replace with accurate data)
continent_mapping = {
    'United Kingdom': 'Europe',
    'France': 'Europe',
    'Australia': 'Oceania',
    'Netherlands': 'Europe',
    'Germany': 'Europe',
    'Norway': 'Europe',
    'EIRE': 'Europe',
    'Switzerland': 'Europe',
    'Spain': 'Europe',
    'Poland': 'Europe',
    'Portugal': 'Europe',
    'Italy': 'Europe',
    'Belgium': 'Europe',
    'Lithuania': 'Europe',
    'Japan': 'Asia',
    'Iceland': 'Europe',
    'Channel Islands': 'Europe',
    'Denmark': 'Europe',
    'Cyprus': 'Europe',
    'Sweden': 'Europe',
    'Austria': 'Europe',
    'Israel': 'Asia',
    'Finland': 'Europe',
    'Bahrain': 'Asia',
    'Greece': 'Europe',
    'Hong Kong': 'Asia',
    'Singapore': 'Asia',
    'Lebanon': 'Asia',
    'United Arab Emirates': 'Asia',
    'Saudi Arabia': 'Asia',
    'Czech Republic': 'Europe',
    'Canada': 'North America',
    'Unspecified': 'Unknown',
    'Brazil': 'South America',
    'USA': 'North America',
    'European Community': 'Europe',
    'Malta': 'Europe',
    'RSA': 'Africa'
}

# Create a DataFrame
data_continent = {'Country': countries, 'Continent': [continent_mapping.get(country, 'Unknown') for country in countries]}
df = pd.DataFrame(data_continent)

# Display the DataFrame
df.head()


Country Continent
0  United Kingdom    Europe
1          France    Europe
2       Australia   Oceania
3     Netherlands    Europe
4         Germany    Europe

In [47]:
df

Country      Continent
0         United Kingdom         Europe
1                 France         Europe
2              Australia        Oceania
3            Netherlands         Europe
4                Germany         Europe
5                 Norway         Europe
6                   EIRE         Europe
7            Switzerland         Europe
8                  Spain         Europe
9                 Poland         Europe
10              Portugal         Europe
11                 Italy         Europe
12               Belgium         Europe
13             Lithuania         Europe
14                 Japan           Asia
15               Iceland         Europe
16       Channel Islands         Europe
17               Denmark         Europe
18                Cyprus         Europe
19                Sweden         Europe
20               Austria         Europe
21                Israel           Asia
22               Finland         Europe
23               Bahrain           Asia
24                Greece         Europe
25             Hong Kong           Asia
26             Singapore           Asia
27               Lebanon           Asia
28  United Arab Emirates           Asia
29          Saudi Arabia           Asia
30        Czech Republic         Europe
31                Canada  North America
32           Unspecified        Unknown
33                Brazil  South America
34                   USA  North America
35    European Community         Europe
36                 Malta         Europe
37                   RSA         Africa

In [48]:
df['Continent'].value_counts()

Europe           24
Asia              8
North America     2
Oceania           1
Unknown           1
South America     1
Africa            1
Name: Continent, dtype: int64

In [49]:
df['Continent'].count()

38

# เช็คประเทศ

In [50]:
df.sort_values(by = 'Continent')

Country      Continent
37                   RSA         Africa
14                 Japan           Asia
29          Saudi Arabia           Asia
28  United Arab Emirates           Asia
27               Lebanon           Asia
26             Singapore           Asia
25             Hong Kong           Asia
23               Bahrain           Asia
21                Israel           Asia
19                Sweden         Europe
20               Austria         Europe
24                Greece         Europe
36                 Malta         Europe
30        Czech Republic         Europe
35    European Community         Europe
22               Finland         Europe
17               Denmark         Europe
0         United Kingdom         Europe
15               Iceland         Europe
1                 France         Europe
3            Netherlands         Europe
4                Germany         Europe
16       Channel Islands         Europe
6                   EIRE         Europe
7            Switzerland         Europe
5                 Norway         Europe
9                 Poland         Europe
10              Portugal         Europe
11                 Italy         Europe
12               Belgium         Europe
13             Lithuania         Europe
8                  Spain         Europe
18                Cyprus         Europe
31                Canada  North America
34                   USA  North America
2              Australia        Oceania
33                Brazil  South America
32           Unspecified        Unknown

In [51]:
data_HW[['date', 'time']] = data['InvoiceDate'].str.split(' ', n=1, expand = True)
data_HW

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
541904  12/09/2011 12:50       0.85     12680.0          France  12/09/2011   
541905  12/09/2011 12:50       2.10     12680.0          France  12/09/2011   
541906  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541907  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541908  12/09/2011 12:50       4.95     12680.0          France  12/09/2011   

         time  
0       08:26  
1       08:26  
2       08:26  
3       08:26  
4       08:26  
...       ...  
541904  12:50  
541905  12:50  
541906  12:50  
541907  12:50  
541908  12:50  

[541909 rows x 10 columns]

In [52]:
data_HW['date'].value_counts()

12/05/2011    5331
12/08/2011    4940
11/29/2011    4313
11/16/2011    4195
11/11/2011    4089
              ... 
3/13/2011      537
12/19/2010     522
05/01/2011     452
12/22/2010     291
02/06/2011     279
Name: date, Length: 305, dtype: int64

In [53]:
data_HW_continent = data_HW.merge(df, how ='left', left_on='Country',right_on='Country')
print(data_HW_continent.shape)
data_HW_continent.head()

(541909, 11)


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country        date   time  \
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010  08:26   
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  08:26   
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010  08:26   
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  08:26   
4  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  08:26   

  Continent  
0    Europe  
1    Europe  
2    Europe  
3    Europe  
4    Europe

In [54]:
data_dadate['date_time'] = pd.to_datetime(data_dadate['InvoiceDate'])

In [55]:
data_dadate.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country           date_time  
0  12/01/2010 08:26       2.55     17850.0  United Kingdom 2010-12-01 08:26:00  
1  12/01/2010 08:26       3.39     17850.0  United Kingdom 2010-12-01 08:26:00  
2  12/01/2010 08:26       2.75     17850.0  United Kingdom 2010-12-01 08:26:00  
3  12/01/2010 08:26       3.39     17850.0  United Kingdom 2010-12-01 08:26:00  
4  12/01/2010 08:26       3.39     17850.0  United Kingdom 2010-12-01 08:26:00

In [56]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [57]:
data.isnull().any()

InvoiceNo      False
StockCode      False
Description     True
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID      True
Country        False
dtype: bool

In [58]:
data[data['Description'].isnull()]

InvoiceNo StockCode Description  Quantity       InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/01/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/01/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/01/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/01/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/01/2010 14:34        0.0   
...          ...       ...         ...       ...               ...        ...   
535322    581199     84581         NaN        -2  12/07/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/07/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/07/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/08/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/08/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

In [62]:
data[data['Description'].isnull()]['StockCode'].value_counts()

35965     10
23084     10
22084      9
22451      6
22501      5
          ..
16169P     1
84846A     1
82545A     1
84251F     1
85175      1
Name: StockCode, Length: 960, dtype: int64

In [59]:
data[data['Quantity']<0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

             InvoiceDate  UnitPrice  CustomerID         Country  
141     12/01/2010 09:41      27.50     14527.0  United Kingdom  
154     12/01/2010 09:49       4.65     15311.0  United Kingdom  
235     12/01/2010 10:24       1.65     17548.0  United Kingdom  
236     12/01/2010 10:24       0.29     17548.0  United Kingdom  
237     12/01/2010 10:24       0.29     17548.0  United Kingdom  
...                  ...        ...         ...             ...  
540449  12/09/2011 09:57       0.83     14397.0  United Kingdom  
541541  12/09/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/09/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/09/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/09/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

In [68]:
data[(data['InvoiceNo'] == 'C581490') | (data['InvoiceNo'] == 'C581499')]

InvoiceNo StockCode                      Description  Quantity  \
540448   C581490     22178  VICTORIAN GLASS HANGING T-LIGHT       -12   
540449   C581490     23144  ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                           Manual        -1   

             InvoiceDate  UnitPrice  CustomerID         Country  
540448  12/09/2011 09:57       1.95     14397.0  United Kingdom  
540449  12/09/2011 09:57       0.83     14397.0  United Kingdom  
541541  12/09/2011 10:28     224.69     15498.0  United Kingdom

In [70]:
data[(data['CustomerID'] == 14397)]

InvoiceNo StockCode                        Description  Quantity  \
182820    552575     22178    VICTORIAN GLASS HANGING T-LIGHT        24   
182821    552575     23144    ZINC T-LIGHT HOLDER STARS SMALL        12   
182822    552575     20846  ZINC HEART LATTICE T-LIGHT HOLDER        12   
182823    552575     82484  WOOD BLACK BOARD ANT WHITE FINISH         2   
182824    552575     84946       ANTIQUE SILVER T-LIGHT GLASS        12   
...          ...       ...                                ...       ...   
533101    581011     22478           BIRDHOUSE GARDEN MARKER         20   
533102    581011     20846  ZINC HEART LATTICE T-LIGHT HOLDER        24   
533103    581011     22600      CHRISTMAS RETROSPOT STAR WOOD        12   
540448   C581490     22178    VICTORIAN GLASS HANGING T-LIGHT       -12   
540449   C581490     23144    ZINC T-LIGHT HOLDER STARS SMALL       -11   

             InvoiceDate  UnitPrice  CustomerID         Country  
182820  05/10/2011 12:25       1.25     14397.0  United Kingdom  
182821  05/10/2011 12:25       0.83     14397.0  United Kingdom  
182822  05/10/2011 12:25       1.25     14397.0  United Kingdom  
182823  05/10/2011 12:25       7.95     14397.0  United Kingdom  
182824  05/10/2011 12:25       1.25     14397.0  United Kingdom  
...                  ...        ...         ...             ...  
533101  12/07/2011 09:22       1.25     14397.0  United Kingdom  
533102  12/07/2011 09:22       1.25     14397.0  United Kingdom  
533103  12/07/2011 09:22       0.85     14397.0  United Kingdom  
540448  12/09/2011 09:57       1.95     14397.0  United Kingdom  
540449  12/09/2011 09:57       0.83     14397.0  United Kingdom  

[95 rows x 8 columns]

In [60]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [71]:
data_dadate

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom   
...                  ...        ...         ...             ...   
541904  12/09/2011 12:50       0.85     12680.0          France   
541905  12/09/2011 12:50       2.10     12680.0          France   
541906  12/09/2011 12:50       4.15     12680.0          France   
541907  12/09/2011 12:50       4.15     12680.0          France   
541908  12/09/2011 12:50       4.95     12680.0          France   

                 date_time  
0      2010-12-01 08:26:00  
1      2010-12-01 08:26:00  
2      2010-12-01 08:26:00  
3      2010-12-01 08:26:00  
4      2010-12-01 08:26:00  
...                    ...  
541904 2011-12-09 12:50:00  
541905 2011-12-09 12:50:00  
541906 2011-12-09 12:50:00  
541907 2011-12-09 12:50:00  
541908 2011-12-09 12:50:00  

[541909 rows x 9 columns]

## test period time

In [75]:
#
# Define the date period
start_month = 12  # April
start_day = 1
end_month = 12    # September
end_day = 31

# Check if each date is within the specified period every year
data_dadate['Within_Period'] = (
    (data_dadate['date_time'].dt.month > start_month) |
    ((data_dadate['date_time'].dt.month == start_month) & (data_dadate['date_time'].dt.day >= start_day))
) & (
    (data_dadate['date_time'].dt.month < end_month) |
    ((data_dadate['date_time'].dt.month == end_month) & (data_dadate['date_time'].dt.day <= end_day))
)

data_dadate

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom   
...                  ...        ...         ...             ...   
541904  12/09/2011 12:50       0.85     12680.0          France   
541905  12/09/2011 12:50       2.10     12680.0          France   
541906  12/09/2011 12:50       4.15     12680.0          France   
541907  12/09/2011 12:50       4.15     12680.0          France   
541908  12/09/2011 12:50       4.95     12680.0          France   

                 date_time  Within_Period  
0      2010-12-01 08:26:00           True  
1      2010-12-01 08:26:00           True  
2      2010-12-01 08:26:00           True  
3      2010-12-01 08:26:00           True  
4      2010-12-01 08:26:00           True  
...                    ...            ...  
541904 2011-12-09 12:50:00           True  
541905 2011-12-09 12:50:00           True  
541906 2011-12-09 12:50:00           True  
541907 2011-12-09 12:50:00           True  
541908 2011-12-09 12:50:00           True  

[541909 rows x 10 columns]

In [78]:
data_xmas = data_dadate[data_dadate['Within_Period']]
data_xmas

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom   
...                  ...        ...         ...             ...   
541904  12/09/2011 12:50       0.85     12680.0          France   
541905  12/09/2011 12:50       2.10     12680.0          France   
541906  12/09/2011 12:50       4.15     12680.0          France   
541907  12/09/2011 12:50       4.15     12680.0          France   
541908  12/09/2011 12:50       4.95     12680.0          France   

                 date_time  Within_Period  
0      2010-12-01 08:26:00           True  
1      2010-12-01 08:26:00           True  
2      2010-12-01 08:26:00           True  
3      2010-12-01 08:26:00           True  
4      2010-12-01 08:26:00           True  
...                    ...            ...  
541904 2011-12-09 12:50:00           True  
541905 2011-12-09 12:50:00           True  
541906 2011-12-09 12:50:00           True  
541907 2011-12-09 12:50:00           True  
541908 2011-12-09 12:50:00           True  

[68006 rows x 10 columns]

#HW1 Q1

In [80]:
transactions_xmas = data_xmas.groupby('InvoiceNo')['StockCode'].apply(list)

In [89]:
min_sup = 0.035

frequentItemsets = list(apriori(transactions_xmas,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22086', '22910'}), support=0.047697368421052634, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.047697368421052634, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])
yes
RelationRecord(items=frozenset({'22835', '22112'}), support=0.036513157894736845, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22835', '22112'}), confidence=0.036513157894736845, lift=1.0), OrderedStatistic(items_base=frozenset({'22112'}), items_add=frozenset({'22835'}), confidence=0.5162790697674419, lift=9.748374981944245), OrderedStatistic(items_base=frozenset({'22835'}), items_add=frozenset({'22112'}), confidence=0.68944099378882, lift=9.748

In [85]:
set(data_xmas[(data_xmas['StockCode']=='22086')]['Description'])

{"PAPER CHAIN KIT 50'S CHRISTMAS "}

In [86]:
set(data_xmas[(data_xmas['StockCode']=='22910')]['Description'])

{'PAPER CHAIN KIT VINTAGE CHRISTMAS'}

In [90]:
set(data_xmas[(data_xmas['StockCode']=='22835')]['Description'])

{'HOT WATER BOTTLE I AM SO POORLY'}

In [91]:
set(data_xmas[(data_xmas['StockCode']=='22112')]['Description'])

{'CHOCOLATE HOT WATER BOTTLE'}

In [93]:
set(data_xmas[(data_xmas['StockCode']=='22726')]['Description'])

{'ALARM CLOCK BAKELIKE GREEN'}

In [92]:
set(data_xmas[(data_xmas['StockCode']=='22727')]['Description'])

{'ALARM CLOCK BAKELIKE RED '}